File: backend/modules/Services/CromaticaCreativa.Modules.Services.Domain/Aggregates/ServiceCategory.cs
using CromaticaCreativa.Modules.Services.Domain.Enums;
using CromaticaCreativa.Modules.Services.Domain.ValueObjects;

namespace CromaticaCreativa.Modules.Services.Domain.Aggregates;

public sealed class ServiceCategory
{
    private ServiceCategory(
        ServiceCategoryId id,
        ServiceId serviceId,
        ServiceCategoryName name,
        string? description,
        MediaReference referenceImage,
        DisplayOrder order)
    {
        Id = id ?? throw new ArgumentNullException(nameof(id));
        ServiceId = serviceId ?? throw new ArgumentNullException(nameof(serviceId));
        Name = name ?? throw new ArgumentNullException(nameof(name));
        Description = description ?? string.Empty;
        ReferenceImage = referenceImage ?? throw new ArgumentNullException(nameof(referenceImage));
        Order = order;
        Status = ServiceCategoryStatus.Inactive;
    }

    public ServiceCategoryId Id { get; }

    public ServiceId ServiceId { get; }

    public ServiceCategoryName Name { get; private set; }

    public string Description { get; private set; }

    public MediaReference ReferenceImage { get; private set; }

    public ServiceCategoryStatus Status { get; private set; }

    public DisplayOrder Order { get; private set; }

    public static ServiceCategory Create(
        ServiceCategoryId id,
        ServiceId serviceId,
        ServiceCategoryName name,
        string? description,
        MediaReference referenceImage,
        DisplayOrder order) =>
        new(id, serviceId, name, description, referenceImage, order);

    public void Rename(ServiceCategoryName name) =>
        Name = name ?? throw new ArgumentNullException(nameof(name));

    public void ChangeDescription(string? description) => Description = description ?? string.Empty;

    public void ChangeReferenceImage(MediaReference referenceImage) =>
        ReferenceImage = referenceImage ?? throw new ArgumentNullException(nameof(referenceImage));

    public void ChangeOrder(DisplayOrder order) => Order = order;

    public void Activate() => Status = ServiceCategoryStatus.Active;

    public void Deactivate() => Status = ServiceCategoryStatus.Inactive;
}

File: backend/src/modules/Services/Services.Domain/Aggregates/ServiceCategory.ts
import { ServiceCategoryStatus } from '../Enums/ServiceCategoryStatus';
import { DisplayOrder } from '../ValueObjects/DisplayOrder';
import { MediaReference } from '../ValueObjects/MediaReference';
import { ServiceCategoryId } from '../ValueObjects/ServiceCategoryId';
import { ServiceCategoryName } from '../ValueObjects/ServiceCategoryName';
import { ServiceId } from '../ValueObjects/ServiceId';

export class ServiceCategory {
  private _status = ServiceCategoryStatus.Inactive;
  private constructor(
    public readonly id: ServiceCategoryId,
    public readonly serviceId: ServiceId,
    private _name: ServiceCategoryName,
    private _description: string,
    private _referenceImage: MediaReference,
    private _order: DisplayOrder,
  ) {}
  public static create(id: ServiceCategoryId, serviceId: ServiceId, name: ServiceCategoryName,
    description: string | null | undefined, referenceImage: MediaReference, order: DisplayOrder): ServiceCategory {
    return new ServiceCategory(id, serviceId, name, description ?? '', referenceImage, order);
  }
  public get name(): ServiceCategoryName { return this._name; }
  public get description(): string { return this._description; }
  public get referenceImage(): MediaReference { return this._referenceImage; }
  public get order(): DisplayOrder { return this._order; }
  public get status(): ServiceCategoryStatus { return this._status; }
  public rename(name: ServiceCategoryName): void { this._name = name; }
  public changeDescription(description?: string | null): void { this._description = description ?? ''; }
  public changeReferenceImage(image: MediaReference): void { this._referenceImage = image; }
  public changeOrder(order: DisplayOrder): void { this._order = order; }
  public activate(): void { this._status = ServiceCategoryStatus.Active; }
  public deactivate(): void { this._status = ServiceCategoryStatus.Inactive; }
}
